File: clients/scala/src/main/scala/komus24/DebugInterface.scala
import java.io.InputStream
import java.io.OutputStream

class DebugInterface(private val inputStream: InputStream, private val outputStream: OutputStream) {
  def addCircle(pos: komus24.model.Vec2Double, radius: Double, color: komus24.debugging.Color): scala.Unit = {
      add(komus24.debugging.DebugData.Circle(pos, radius, color))
  }
  
  def addLine(point1: komus24.model.Vec2Double, point2: komus24.model.Vec2Double, width: Double, color: komus24.debugging.Color): scala.Unit = {
      add(komus24.debugging.DebugData.Line(point1, point2, width, color))
  }
  
  def addRect(corner1: komus24.model.Vec2Double, corner2: komus24.model.Vec2Double, color: komus24.debugging.Color): scala.Unit = {
      add(komus24.debugging.DebugData.Rect(corner1, corner2, color))
  }
  
  def addText(text: String, pos: komus24.model.Vec2Double, size: Double, align: Double, color: komus24.debugging.Color): scala.Unit = {
      add(komus24.debugging.DebugData.Text(text, pos, size, align, color))
  }

  def add(debugData: komus24.debugging.DebugData): scala.Unit = {
      send(komus24.debugging.DebugCommand.Add(debugData))
  }
  
  def clear(): scala.Unit = {
      send(komus24.debugging.DebugCommand.Clear())
  }
  
  def setAutoFlush(enable: Boolean): scala.Unit = {
      send(komus24.debugging.DebugCommand.SetAutoFlush(enable))
  }
  
  def flush(): scala.Unit = {
      send(komus24.debugging.DebugCommand.Flush())
  }

  def send(command: komus24.debugging.DebugCommand): scala.Unit = {
    komus24.codegame.ClientMessage.DebugMessage(command).writeTo(outputStream)
    outputStream.flush()
  }

  def getState(): komus24.debugging.DebugState = {
    komus24.codegame.ClientMessage.RequestDebugState().writeTo(outputStream)
    outputStream.flush()
    komus24.debugging.DebugState.readFrom(inputStream)
  }
}

File: clients/swift/Sources/Komus24/DebugInterface.swift
class DebugInterface {
    private let inputStream: BufferedInputStream
    private let outputStream: BufferedOutputStream

    init(_ inputStream: BufferedInputStream, _ outputStream: BufferedOutputStream) {
        self.inputStream = inputStream
        self.outputStream = outputStream
    }

    func send(_ command: DebugCommand) {
        ClientMessage.debugMessage(command: command).writeTo(outputStream)
        outputStream.flush()
    }

    func getState() -> DebugState {
        ClientMessage.requestDebugState.writeTo(outputStream)
        outputStream.flush()
        return DebugState.readFrom(inputStream)
    }

    func addCircle(_ pos: Vec2Double, _ radius: Double, _ color: Color) {
        self.add(DebugData.circle(pos: pos, radius: radius, color: color))
    }
    func addLine(_ point1: Vec2Double, _ point2: Vec2Double, _ width: Double, _ color: Color) {
        self.add(DebugData.line(point1: point1, point2: point2, width: width, color: color))
    }
    func addRect(_ corner1: Vec2Double, _ corner2: Vec2Double, _ color: Color) {
        self.add(DebugData.rect(corner1: corner1, corner2: corner2, color: color))
    }
    func addText(_ text: String, _ pos: Vec2Double, _ size: Double, _ align: Double, _ color: Color) {
        self.add(DebugData.text(text: text, pos: pos, size: size, align: align, color: color))
    }

    func add(_ debugData: DebugData) {
        self.send(DebugCommand.add(debugData: debugData))
    }
    func clear() {
        self.send(DebugCommand.clear)
    }
    func setAutoFlush(_ enable: Bool) {
        self.send(DebugCommand.setAutoFlush(enable: enable))
    }
    func flush() {
        self.send(DebugCommand.flush)
    }
}